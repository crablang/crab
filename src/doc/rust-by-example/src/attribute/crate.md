# Crates

The `crate_type` attribute can be used to tell the compiler whether a crate is
a binary or a library (and even which type of library), and the `crate_name`
attribute can be used to set the name of the crate.

However, it is important to note that both the `crate_type` and `crate_name`
attributes have **no** effect whatsoever when using Cargo, the Rust package
manager. Since Cargo is used for the majority of Rust projects, this means
real-world uses of `crate_type` and `crate_name` are relatively limited.

```rust,editable
// This crate is a library
#![crate_type = "lib"]
// The library is named "rary"
#![crate_name = "rary"]

pub fn public_function() {
    println!("called rary's `public_function()`");
}

fn private_function() {
    println!("called rary's `private_function()`");
}

pub fn indirect_access() {
    print!("called rary's `indirect_access()`, that\n> ");

    private_function();
}
```

When the `crate_type` attribute is used, we no longer need to pass the
`--crate-type` flag to `rustc`.

```shell
$ rustc lib.rs
$ ls lib*
library.rlib
```
