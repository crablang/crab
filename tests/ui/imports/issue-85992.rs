// edition: 2021
// compile-flags: --extern issue_85992_extern_1 --extern issue_85992_extern_2
// aux-build: issue-85992-extern-1.rs
// aux-build: issue-85992-extern-2.rs

issue_85992_extern_1::m!();

use crate::issue_85992_extern_2;
//~^ ERROR unresolved import `crate::issue_85992_extern_2`

fn main() {}
