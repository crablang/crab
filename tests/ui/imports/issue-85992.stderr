error[E0432]: unresolved import `crate::issue_85992_extern_2`
  --> $DIR/issue-85992.rs:8:5
   |
LL | use crate::issue_85992_extern_2;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `issue_85992_extern_2` in the root

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
